dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Definition.java

```diff
@@ -0,0 +1,106 @@
+        default T visitDifference(Relation rel, Relation superset, Relation complement) { return visitDefinition(rel, List.of(superset, complement)); }
+        default T visitComposition(Relation rel, Relation front, Relation back) { return visitDefinition(rel, List.of(front, back)); }
+        default T visitDomainIdentity(Relation rel, Relation operand) { return visitDefinition(rel, List.of(operand)); }
+        default T visitRangeIdentity(Relation rel, Relation operand) { return visitDefinition(rel, List.of(operand)); }
+        default T visitInverse(Relation rel, Relation operand) { return visitDefinition(rel, List.of(operand)); }
+        default T visitTransitiveClosure(Relation rel, Relation operand) { return visitDefinition(rel, List.of(operand)); }
+        default T visitEmpty(Relation rel) { return visitDefinition(rel, List.of()); }
+        default T visitIdentity(Relation id, FilterAbstract set) { return visitDefinition(id, List.of()); }
+        default T visitProduct(Relation rel, FilterAbstract domain, FilterAbstract range) { return visitDefinition(rel, List.of()); }
+        default T visitExternal(Relation ext) { return visitDefinition(ext, List.of()); }
+        default T visitInternal(Relation int_) { return visitDefinition(int_, List.of()); }
+        default T visitProgramOrder(Relation po, FilterAbstract type) { return visitDefinition(po, List.of()); }
+        default T visitControl(Relation ctrlDirect) { return visitDefinition(ctrlDirect, List.of()); }
+        default T visitFences(Relation fence, FilterAbstract type) { return visitDefinition(fence, List.of()); }
+        default T visitInternalDataDependency(Relation idd) { return visitDefinition(idd, List.of()); }
+        default T visitCompareAndSwapDependency(Relation casDep) { return visitDefinition(casDep, List.of()); }
+        default T visitAddressDependency(Relation addrDirect) { return visitDefinition(addrDirect, List.of()); }
+        default T visitCriticalSections(Relation rscs) { return visitDefinition(rscs, List.of()); }
+        default T visitReadModifyWrites(Relation rmw) { return visitDefinition(rmw, List.of()); }
+        default T visitMemoryOrder(Relation co) { return visitDefinition(co, List.of()); }
+        default T visitSameAddress(Relation loc) { return visitDefinition(loc, List.of()); }
+        default T visitReadFrom(Relation rf) { return visitDefinition(rf, List.of()); }
+    }
+
+    private String getTerm(int depth) {
+        List<Relation> l = getConstrainedRelations();
+        int s = l.size() - 1;
+        Object[] o = new Object[s];
+        for (int i = 0; i < s; i++) {
+            Relation r = l.get(i + 1);
+            o[i] = depth == 0 ? r.definition.getTerm(depth - 1) : r.getName() != null ? r.getName() : "(" + r.definition.getTerm(depth - 1) + ")";
+        }
+        return String.format(term, o);
```

Comment:
Something is wrong here ` o[i] = depth == 0 ? r.definition.getTerm(depth - 1) : ...`. It recursively descends if the depth is 0.
Also, rather than tracking some depth to unroll recursive definitions a bounded number of times, I think it might be better to use a `Set<Definition>` to track which definitions are recursively visited and place some placeholder name there to indicate that the definition can be unfolded there an unlimited amount of times.

Reply:
That's indeed a bug.  I think a fixed-point algorithm is immoderate in this case.  Unnamed recursive definitions still should not occur in this project, but if they do, diagnostics should not fail.

Reply:
The "fixed-point" algorithm is simpler than arbitrarily bounding the depth, no?
I would just do something like `getTermRecursive(Set<Definition> alreadyVisited)` and then define `getTerm() = getTermRecursive(new HashSet())`.
The check whether to "descend" then changes from `depth == 0` to `alreadyVisited.contains(...)`.
I would say the code amount stays the same but the logic gets more straightforward and less arbitrary (the depth bound is kinda arbitrary).